source/slang/slang-ir-lower-l-value-cast.cpp

```diff
@@ -0,0 +1,300 @@
+
+            if (useSites.indexOf(useSite) < 0)
+            {
+                useSites.add(useSite);
+            }
+        }
+
+        auto nameHintDecoration = castOperand->findDecoration<IRNameHintDecoration>();
+
+        IRBuilder builder(m_module);
+        
+        IRType* toValueType = as<IRPtrType>(toType)->getValueType();
+        IRType* fromValueType = as<IRPtrType>(fromType)->getValueType();
+
+        for (auto useSite : useSites)
+        {
+            builder.setInsertBefore(useSite);
+            auto tmpVar = builder.emitVar(toValueType);
+
+            if (nameHintDecoration)
+            {
+                cloneDecoration(nameHintDecoration, tmpVar);
+            }
+
+            // If it's inout we convert via cast whats in the variable
+            if (castInst->getOp() == kIROp_InOutImplicitCast)
+            {
+                builder.emitStore(tmpVar, builder.emitCast(toValueType, builder.emitLoad(castOperand)));
+            }
+
+            // Convert the temporary back to the original location
+            builder.setInsertAfter(useSite);
+            builder.emitStore(castOperand, builder.emitCast(fromValueType, builder.emitLoad(tmpVar)));
+
+            // Go through all of the operands of the inst relacing 
+
+            const auto operandCount = Count(useSite->getOperandCount());
+            auto operands = useSite->getOperands();
+
+            for (Index i = 0; i < operandCount; ++i)
+            {
+                auto& callSiteOperand = operands[i];
+                if(callSiteOperand.get() == castInst)
+                {
+                    callSiteOperand.set(tmpVar);
+                }
+            }
+        }
+
+        // When we are done we can destroy the inst
+        castInst->removeAndDeallocate();
+    }
+
+
+    LValueCastLoweringContext(TargetRequest* targetRequest, IRModule* module):
+        m_targetReq(targetRequest),
+        m_module(module)
+    {
+        m_sourceLanguage = _calcIntermediateSourceLanguage(targetRequest);
+    }
+
+    // TODO(JS):
+    // This would probably be better served elsewhere, as it bakes in the knowledge of *how* 
+    // code generate is performed which could change
+    static SourceLanguage _calcIntermediateSourceLanguage(TargetRequest* req)
+    {
+        // If we are emitting directly just do the default thing
+        if (req->shouldEmitSPIRVDirectly())
+        {
+            return SourceLanguage::Unknown;
+        }
+
+        switch (req->getTarget())
+        {
+            case CodeGenTarget::GLSL:
+            case CodeGenTarget::GLSL_Vulkan:
+            case CodeGenTarget::GLSL_Vulkan_OneDesc:
+            // If we aren't emitting directly we are going to output GLSL to feed to GLSLANG
+            case CodeGenTarget::SPIRV:
+            case CodeGenTarget::SPIRVAssembly:
+            {
+                return SourceLanguage::GLSL;
+            }
+            case CodeGenTarget::HLSL:
+            case CodeGenTarget::DXBytecode:
```

Comment:
We already have `isKhronosTarget` `isD3DTarget` `isCUDATarget` and `isCPUTarget` utility functions defined. We can just use those, or move this function to sit alongside with those?

Reply:
Moved to same location as isKronos etc.